src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -257,38 +257,42 @@ private SparseFixedBitSet bitSetForSearch()
     }
 
     @Override
-    public RangeIterator<PrimaryKey> limitToTopResults(QueryContext context, RangeIterator<Long> iterator, Expression exp, int limit) throws IOException
+    public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> keys, Expression exp, int limit) throws IOException
     {
+        // VSTODO would it be better to do a binary search to find the boundaries?
+        List<PrimaryKey> keysInRange = keys.stream()
+                                           .dropWhile(k -> k.compareTo(metadata.minKey) < 0)
+                                           .takeWhile(k -> k.compareTo(metadata.maxKey) <= 0)
+                                           .collect(Collectors.toList());
         try (PrimaryKeyMap primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap())
         {
-            // the iterator represents keys from all the segments in our sstable -- we'll only pull of those that
+            // the iterator represents keys from the whole table -- we'll only pull of those that
             // are from our own token range so we can use row ids to order the results by vector similarity.
             var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
             SparseFixedBitSet bits = bitSetForSearch();
             var rowIds = new IntArrayList();
             try (var ordinalsView = graph.getOrdinalsView())
             {
-                while (iterator.hasNext())
+                for (PrimaryKey primaryKey : keysInRange)
                 {
-                    Long sstableRowId = iterator.peek();
-                    // if sstable row id has exceeded current ANN segment, stop
-                    if (sstableRowId > metadata.maxSSTableRowId)
-                        break;
-
-                    iterator.next();
+                    // VSTODO this is O(log n), is this a hotspot?
+                    long sstableRowId = primaryKeyMap.exactRowIdForPrimaryKey(primaryKey);
                     // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
+                    // or are not in this segment (exactRowIdForPrimaryKey returns a negative value for not found)
                     if (sstableRowId < metadata.minSSTableRowId)
                         continue;
 
+                    // if sstable row id has exceeded current ANN segment, stop
+                    if (sstableRowId > metadata.maxSSTableRowId)
+                        break;
+
                     int segmentRowId = metadata.toSegmentRowId(sstableRowId);
                     rowIds.add(segmentRowId);
-
+                    // VSTODO now that we know the size of keys evaluated, is it worth doing the brute
```

Comment:
we should do this

this means we'd be doing two checks, right?  because we have an upper bound before doing the pk map, but it could shrink based on the segment checks

tbc i think that's fine, the brute force check is not expensive

Reply:
Yes, now that we've accepted using `exactRowIdForPrimaryKey`, we should do it.

I also noticed another minor improvement would pre-filter the list of PKs based on the segment boundaries so that we don't iterate each list within every segment.

Reply:
This is done. I did not do the pre-batching yet because the segment class is missing the right methods. I'll follow up later.